File: service-database/src/main/java/com/bnorm/barkeep/db/DbRecipeService.java
// Copyright 2016 (C) BNORM Software. All rights reserved.
package com.bnorm.barkeep.db;

import java.util.List;

import javax.persistence.EntityManager;

import com.bnorm.barkeep.model.Component;
import com.bnorm.barkeep.model.Recipe;
import com.bnorm.barkeep.service.RecipeService;

public class DbRecipeService extends AbstractDbService implements RecipeService {

  public DbRecipeService(EntityManager entityManager) {
    super(entityManager);
  }

  @Override
  public List<Recipe> getRecipes() {
    return super.getRecipes();
  }

  @Override
  public RecipeEntity getRecipe(long recipeId) {
    return super.getRecipe(recipeId);
  }

  @Override
  public RecipeEntity createRecipe(Recipe recipe) {
    if (recipe.getId() != null) {
      throw new IllegalArgumentException(String.format("Cannot create recipe that already has an id=%d",
                                                       recipe.getId()));
    } else if (recipe.getOwner() == null) {
      throw new IllegalArgumentException("Cannot create recipe without an owner");
    }
    UserEntity userEntity = find(recipe.getOwner());
    if (userEntity == null) {
      throw new IllegalArgumentException(String.format("Cannot create recipe with an unknown owner id=%d",
                                                       recipe.getOwner().getId()));
    }

    return transaction(em -> {
      RecipeEntity recipeEntity = new RecipeEntity();

      recipeEntity.setTitle(recipe.getTitle());
      recipeEntity.setDescription(recipe.getDescription());
      recipeEntity.setOwner(userEntity);
      recipeEntity.setImageUrl(recipe.getImageUrl());
      recipeEntity.setInstructions(recipe.getInstructions());
      recipeEntity.setSource(recipe.getInstructions());

      if (recipe.getComponents() != null) {
        for (Component component : recipe.getComponents()) {
          ComponentEntity componentEntity = new ComponentEntity();
          componentEntity.setIngredient(find(component.getIngredient()));
          componentEntity.setMin(component.getMin());
          componentEntity.setMax(component.getMax());
          componentEntity.setOrder(component.getOrder());
          componentEntity.setComponentNum(component.getComponentNum());
          recipeEntity.addComponent(componentEntity);
        }
      }

      em.persist(recipeEntity);

      userEntity.addRecipe(recipeEntity);
      return recipeEntity;
    });
  }

  @Override
  public RecipeEntity setRecipe(long recipeId, Recipe recipe) {
    RecipeEntity recipeEntity = findRecipe(recipeId);
    requireExists(recipeEntity, recipeId, "recipe");
    requireMatch(recipe, recipeId, "recipe");

    return transaction(em -> {
      if (recipe.getTitle() != null) {
        recipeEntity.setTitle(recipe.getTitle());
      }
      if (recipe.getDescription() != null) {
        recipeEntity.setDescription(recipe.getTitle());
      }
      if (recipe.getImageUrl() != null) {
        recipeEntity.setImageUrl(recipe.getImageUrl());
      }
      if (recipe.getInstructions() != null) {
        recipeEntity.setInstructions(recipe.getInstructions());
      }
      if (recipe.getSource() != null) {
        recipeEntity.setSource(recipe.getSource());
      }
      if (recipe.getComponents() != null) {
        // todo is this right?
        for (Component component : recipe.getComponents()) {
          ComponentEntity componentEntity = new ComponentEntity();
          componentEntity.setIngredient(find(component.getIngredient()));
          componentEntity.setMin(component.getMin());
          componentEntity.setMax(component.getMax());
          componentEntity.setOrder(component.getOrder());
          componentEntity.setComponentNum(component.getComponentNum());
          recipeEntity.addComponent(componentEntity);
        }
      }

      return recipeEntity;
    });
  }

  @Override
  public void deleteRecipe(long recipeId) {
    transaction(em -> {
      RecipeEntity recipeEntity = findRecipe(recipeId);
      requireExists(recipeEntity, recipeId, "recipe");
      em.remove(recipeEntity);
    });
  }
}

File: service-database/src/main/kotlin/com/bnorm/barkeep/db/DbRecipeService.kt
// Copyright 2017 (C) BNORM Software. All rights reserved.
package com.bnorm.barkeep.db

import com.bnorm.barkeep.model.Recipe
import com.bnorm.barkeep.service.RecipeService
import javax.persistence.EntityManager

class DbRecipeService(entityManager: EntityManager) : AbstractDbService(entityManager), RecipeService {

  override fun getRecipes(): List<Recipe> {
    return super.getRecipes()
  }

  override fun getRecipe(id: Long): RecipeEntity? {
    return super.getRecipe(id)
  }

  override fun createRecipe(recipe: Recipe): RecipeEntity {
    if (recipe.id != null) {
      throw IllegalArgumentException("Cannot create recipe that already has an id=$recipe.id")
    } else if (recipe.owner == null) {
      throw IllegalArgumentException("Cannot create recipe without an owner")
    }
    val userEntity = find(recipe.owner!!) ?: throw IllegalArgumentException("Cannot create recipe with an unknown owner id=$recipe.owner!!.id")

    val recipeEntity = RecipeEntity()

    recipeEntity.title = recipe.title
    recipeEntity.description = recipe.description
    recipeEntity.owner = userEntity
    recipeEntity.imageUrl = recipe.imageUrl
    recipeEntity.instructions = recipe.instructions
    recipeEntity.source = recipe.instructions

    if (recipe.components != null) {
      for (component in recipe.components!!) {
        val componentEntity = ComponentEntity()
        componentEntity.ingredient = find(component.ingredient!!)
        componentEntity.min = component.min
        componentEntity.max = component.max
        componentEntity.order = component.order
        componentEntity.componentNum = component.componentNum
        recipeEntity.addComponent(componentEntity)
      }
    }

    txn {
      em.persist(recipeEntity)
      userEntity.addRecipe(recipeEntity)
    }

    return recipeEntity
  }

  override fun setRecipe(id: Long, recipe: Recipe): RecipeEntity {
    val recipeEntity = requireExists(findRecipe(id), id, "recipe")
    requireMatch(recipe, id, "recipe")

    txn {
      if (recipe.title != null) {
        recipeEntity.title = recipe.title
      }
      if (recipe.description != null) {
        recipeEntity.description = recipe.title
      }
      if (recipe.imageUrl != null) {
        recipeEntity.imageUrl = recipe.imageUrl
      }
      if (recipe.instructions != null) {
        recipeEntity.instructions = recipe.instructions
      }
      if (recipe.source != null) {
        recipeEntity.source = recipe.source
      }
      if (recipe.components != null) {
        // todo is this right?
        for (component in recipe.components!!) {
          val componentEntity = ComponentEntity()
          componentEntity.ingredient = find(component.ingredient!!)
          componentEntity.min = component.min
          componentEntity.max = component.max
          componentEntity.order = component.order
          componentEntity.componentNum = component.componentNum
          recipeEntity.addComponent(componentEntity)
        }
      }

    }
    return recipeEntity
  }

  override fun deleteRecipe(id: Long) {
    txn {
      val recipeEntity = findRecipe(id)
      requireExists(recipeEntity, id, "recipe")
      em.remove(recipeEntity)
    }
  }
}
